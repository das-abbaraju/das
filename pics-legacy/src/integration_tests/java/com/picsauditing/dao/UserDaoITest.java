package com.picsauditing.dao;

import static org.junit.Assert.*;
import static org.hamcrest.Matchers.*;

import com.picsauditing.PICS.I18nCache;
import com.picsauditing.jpa.entities.User;
import com.picsauditing.search.Database;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.reflect.Whitebox;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration(locations={"UserDAOITest-context.xml"})
public class UserDaoITest {
    @Mock
    private Database databaseForTesting;
    @Autowired
    private UserDAO userDAO;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        Whitebox.setInternalState(I18nCache.class, "databaseForTesting", databaseForTesting);
    }


    @Test
    public void testFindName() throws Exception {
        User user = userDAO.findName("gmeurer");
        assertThat(user, is(notNullValue()));
    }
}
