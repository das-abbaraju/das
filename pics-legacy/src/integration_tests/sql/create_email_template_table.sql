DROP TABLE IF EXISTS `email_template`;

CREATE TABLE `email_template` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `accountID` mediumint(8) unsigned NOT NULL,
  `templateName` varchar(50) NOT NULL,
  `subject` varchar(150) DEFAULT NULL,
  `body` text,
  `createdBy` int(11) DEFAULT NULL,
  `creationDate` datetime DEFAULT NULL,
  `updatedBy` int(11) DEFAULT NULL,
  `updateDate` datetime DEFAULT NULL,
  `listType` varchar(50) DEFAULT 'Contractor',
  `allowsVelocity` tinyint(4) NOT NULL DEFAULT '0',
  `html` tinyint(4) NOT NULL DEFAULT '0',
  `recipient` varchar(10) DEFAULT NULL,
  `translated` tinyint(4) NOT NULL DEFAULT '0',
  `requiredLanguages` varchar(100) DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `accountID` (`accountID`,`templateName`,`listType`)
) ENGINE=InnoDB AUTO_INCREMENT=234 DEFAULT CHARSET=latin1;