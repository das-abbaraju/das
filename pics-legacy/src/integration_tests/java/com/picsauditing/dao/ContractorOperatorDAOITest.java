package com.picsauditing.dao;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import javax.sql.DataSource;

import com.picsauditing.jpa.entities.ContractorAccount;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.powermock.reflect.Whitebox;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import com.picsauditing.PICS.I18nCache;
import com.picsauditing.jpa.entities.Account;
import com.picsauditing.jpa.entities.ContractorOperator;
import com.picsauditing.search.Database;

import static org.junit.Assert.assertTrue;

@RunWith(SpringJUnit4ClassRunner.class)
@ContextConfiguration(locations={"ContractorOperatorDAOITest-context.xml"})
public class ContractorOperatorDAOITest {
	private JdbcTemplate jdbcTemplate;
	
	@Autowired private ContractorOperatorDAO contractorOperatorDAO;
	@Autowired private ContractorAccountDAO contractorAccountDAO;

	@Mock private Database databaseForTesting;
	
	@Autowired 
    public void setDataSource(DataSource dataSource) {
        this.jdbcTemplate = new JdbcTemplate(dataSource);
    }
	
	@Before
	public void setUp() throws Exception {
		MockitoAnnotations.initMocks(this);
		Whitebox.setInternalState(I18nCache.class, "databaseForTesting", databaseForTesting);
	}

    @Test
    public void testCA() {
        ContractorAccount contractorAccount = contractorAccountDAO.find(39202);
        assertTrue(true);
    }

	@Test
	public void testFindForOperators() {
		@SuppressWarnings({ "rawtypes", "unchecked", "serial" })
		Set<Integer> operatorIds = new HashSet() {{
			add(16); add(1068); add(1251); 
		}};
		Set<ContractorOperator> results = contractorOperatorDAO.findForOperators(357, operatorIds);
	}
}
