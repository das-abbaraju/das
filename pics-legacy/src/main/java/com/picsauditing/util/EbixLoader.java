package com.picsauditing.util;

import com.picsauditing.actions.cron.CronTask;
import com.picsauditing.actions.cron.CronTaskException;
import com.picsauditing.actions.cron.CronTaskResult;
import com.picsauditing.dao.AppPropertyDAO;
import com.picsauditing.dao.ContractorAccountDAO;
import com.picsauditing.dao.ContractorAuditDAO;
import com.picsauditing.jpa.entities.*;
import org.apache.commons.net.ftp.FTPClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.util.CollectionUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.List;

public class EbixLoader implements CronTask {
    @Autowired
    AppPropertyDAO appPropDao;
    @Autowired
    ContractorAuditDAO contractorAuditDAO;
    @Autowired
    ContractorAccountDAO contractorAccountDAO;

    private final Logger logger = LoggerFactory.getLogger(EbixLoader.class);

    public String getDescription() {
        return "Update Huntsman FTP";
    }

    public List<String> getSteps() {
        return null;
    }

    public CronTaskResult run() {
        CronTaskResult results = new CronTaskResult();
        String server = appPropDao.find("huntsmansync.ftp.server").getValue();
        String username = appPropDao.find("huntsmansync.ftp.user").getValue();
        String password = appPropDao.find("huntsmansync.ftp.password").getValue();
        String folder = appPropDao.find("huntsmansync.ftp.folder").getValue();

        logger.debug("Server: " + server);
        logger.debug("username: " + username);
        logger.debug("folder: " + folder);

        FTPClient ftp = new FTPClient();

        logger.debug("logging in to server...");

        try {
            ftp.connect(server);
            ftp.enterLocalPassiveMode();

            ftp.login(username, password);

            ftp.changeWorkingDirectory(folder);

            String[] names = ftp.listNames();

            if (names != null) {
                for (String fileName : names) {
                    processFile(ftp, fileName);
                }
            }

            ftp.logout();
        } catch (IOException e) {

        } finally {
            try {
                ftp.disconnect();
            } catch (IOException e) {
                logger.error("Failed to disconnect FTP because - " + e.getMessage());
            }
        }
        return results;
    }

    private void processFile(FTPClient ftp, String fileName) throws IOException {
        logger.debug("Processing file: " + fileName);

        BufferedReader reader = null;

        InputStream retrieveFileStream = ftp.retrieveFileStream(fileName);

        if (retrieveFileStream != null) {

            reader = new BufferedReader(new InputStreamReader(retrieveFileStream));

            String line = null;

            while ((line = reader.readLine()) != null) {

                if (line.length() > 0) {

                    String[] data = line.split(",");
                    logger.debug("Processing data: " + data[0] + "/" + data[1]);

                    int contractorId = 0;
                    try {
                        contractorId = Integer.parseInt(data[0]);
                    } catch (Exception ignoreStrings) {
                        // Sometimes we get ids that are strings like
                        // HC00000629
                    }

                    if (data.length == 2 && contractorId > 0) {

                        // the other field. comes in as a Y/N.
                        AuditStatus status = AuditStatus.Pending;
                        if (data[1].equals("Y"))
                            status = AuditStatus.Complete;

                        try {
                            List<ContractorAudit> audits = null;
                            ContractorAccount conAccount = contractorAccountDAO.find(contractorId);
                            if (conAccount != null) {
                                audits = contractorAuditDAO.findWhere(900, "auditType.id = "
                                        + AuditType.HUNTSMAN_EBIX + " and contractorAccount.id = "
                                        + conAccount.getId(), "");
                            }

                            if (CollectionUtils.isEmpty(audits)) {
                                if (conAccount != null) {
                                    logger.warn("WARNING: Ebix record found for contractor "
                                            + conAccount.getId() + " but no Ebix Compliance audit was found");
                                } else {
                                    logger.warn("WARNING: Ebix record found for contractor MISSING CONTRACTOR ID but no Ebix Compliance audit was found");
                                }
                                continue;
                            }

                            for (ContractorAudit audit : audits) {
                                logger.debug("Setting Ebix audit " + audit.getId() + " for contractor "
                                        + conAccount.getId() + " to " + status.name());
                                for (ContractorAuditOperator cao : audit.getOperators()) {
                                    if (status != cao.getStatus()) {
                                        cao.changeStatus(status, null);
                                        contractorAuditDAO.save(audit);

                                        conAccount.incrementRecalculation();
                                        contractorAccountDAO.save(conAccount);
                                    } else {
                                        logger.debug("No change for Ebix audit " + audit.getId()
                                                + " for contractor " + conAccount.getId() + ", "
                                                + status.name());
                                    }
                                }
                            }

                        } catch (Exception e) {
                            logger.error("ERROR: Error Processing Ebix for contractor " + data[0]);
                            e.printStackTrace();
                        }

                    } else {
                        logger.error("Bad Data Found : " + data);
                    }
                }
            }
        } else {
            logger.error("unable to open connection: " + ftp.getReplyCode() + ":" + ftp.getReplyString());
        }
    }
}
