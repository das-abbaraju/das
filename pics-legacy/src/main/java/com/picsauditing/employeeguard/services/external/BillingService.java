package com.picsauditing.employeeguard.services.external;

import com.picsauditing.jpa.entities.Account;
import com.picsauditing.jpa.entities.ContractorAccount;
import com.picsauditing.jpa.entities.OperatorAccount;

import java.util.Iterator;
import java.util.List;

/**
 * Stand in for billing logic
 */
public class BillingService {

	public <E extends Account> List<E> filterEmployeeGUARDAccounts(final List<E> accounts) {
		Iterator<E> iterator = accounts.iterator();

		while (iterator.hasNext()) {
			E account = iterator.next();

			if (account.isOperator() && !operatorHasEmployeeGUARD(account)) {
				iterator.remove();
			}

			if (account.isContractor() && !contractorHasEmployeeGUARD(account)) {
				iterator.remove();
			}
		}

		return accounts;
	}

	private <E extends Account> boolean operatorHasEmployeeGUARD(final E account) {
		return ((OperatorAccount) account).isRequiresEmployeeGuard();
	}

	/**
	 * Waiting on requirements for how to bill contractors with EmployeeGUARD operators
	 *
	 * @param account
	 * @param <E>
	 * @return
	 */
	private <E extends Account> boolean contractorHasEmployeeGUARD(final E account) {
		return ((ContractorAccount) account).isHasEmployeeGuard();
	}
}
