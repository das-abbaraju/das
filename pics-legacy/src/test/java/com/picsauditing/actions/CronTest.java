package com.picsauditing.actions;

import com.picsauditing.EntityFactory;
import com.picsauditing.PICS.BillingService;
import com.picsauditing.PICS.DBBean;
import com.picsauditing.PicsActionTest;
import com.picsauditing.PicsTestUtil;
import com.picsauditing.access.Permissions;
import com.picsauditing.dao.*;
import com.picsauditing.jpa.entities.*;
import com.picsauditing.mail.EmailBuilder;
import com.picsauditing.mail.EmailSender;
import com.picsauditing.model.account.AccountStatusChanges;
import com.picsauditing.search.Database;
import com.picsauditing.util.EbixLoader;
import com.picsauditing.util.IndexerEngine;
import com.picsauditing.util.SapAppPropertyUtil;
import org.apache.commons.beanutils.BasicDynaBean;
import org.approvaltests.Approvals;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.UseReporter;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import org.powermock.reflect.Whitebox;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.*;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.*;

@SuppressWarnings("deprecation")
@UseReporter(DiffReporter.class)
public class CronTest extends PicsActionTest {
    private Cron cron;
    private ContractorAccount contractor;
    private OperatorAccount operator;
    private OperatorAccount anotherOperator;
    private List<ContractorOperator> operators;
    @Mock
    private BasicDAO dao;
    @Mock
    private ContractorAccountDAO contractorAccountDAO;
    @Mock
    private ContractorAuditDAO contractorAuditDAO;
    @Mock
    private ContractorAuditOperatorDAO contractorAuditOperatorDAO;
    @Mock
    private ContractorOperatorDAO contractorOperatorDAO;
    @Mock
    private ContractorRegistrationRequestDAO contractorRegistrationRequestDAO;
    @Mock
    private EmailQueueDAO emailQueueDAO;
    @Mock
    private EbixLoader ebixLoader;
    @Mock
    private IndexerEngine indexer;
    @Mock
    private FlagDataOverrideDAO flagDataOverrideDAO;
    @Mock
    private InvoiceDAO invoiceDAO;
    @Mock
    private InvoiceFeeDAO invoiceFeeDAO;
    @Mock
    private InvoiceItemDAO invoiceItemDAO;
    @Mock
    private OperatorAccountDAO operatorDAO;
    @Mock
    private NoteDAO noteDAO;
    @Mock
    private UserDAO userDAO;
    @Mock
    private EmailBuilder emailBuilder;
    @Mock
    private AccountStatusChanges accountStatusChanges;
    @Mock
    private EmailSender emailSender;
    @Mock
    private BillingService billingService;
    @Mock
    private Invoice mockInvoice;
    @Mock
    private InvoiceItem mockInvoiceItem;
    @Mock
    private ContractorAccount mockContractorAccount;
    @Mock
    private InvoiceFee mockInvoiceFee;
    @Mock
    private SapAppPropertyUtil mockSapAppPropertyUtil;
    @Mock
    private Database database;
    @Mock
    private ReportDAO reportDAO;

    @BeforeClass
    public static void classSetUp() {
        Whitebox.setInternalState(DBBean.class, "staticDataSource", Mockito.mock(DataSource.class));
    }

    @AfterClass
    public static void classTearDown() {
        Whitebox.setInternalState(DBBean.class, "staticDataSource", (DataSource) null);
    }

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);


        cron = new Cron();
        super.setUp(cron);

        PicsTestUtil.autowireDAOsFromDeclaredMocks(cron, this);
        Whitebox.setInternalState(cron, "dao", dao);
        Whitebox.setInternalState(cron, "emailBuilder", emailBuilder);
        Whitebox.setInternalState(cron, "emailSender", emailSender);
        Whitebox.setInternalState(cron, "accountStatusChanges", accountStatusChanges);
        Whitebox.setInternalState(cron, "billingService", billingService);
        Whitebox.setInternalState(cron, "ebixLoader", ebixLoader);
        Whitebox.setInternalState(cron, "indexer", indexer);
        Whitebox.setInternalState(cron, "invoiceFeeDAO", invoiceFeeDAO);
        Whitebox.setInternalState(cron, "reportDAO", reportDAO);
        cron.database = this.database;

        operators = new ArrayList<ContractorOperator>();
        // certList = new ArrayList<Certificate>();
        // opIdsByCertIds = new HashMap<Integer, List<Integer>>();
        contractor = EntityFactory.makeContractor();
        operator = EntityFactory.makeOperator();
        anotherOperator = EntityFactory.makeOperator();
        operators.add(EntityFactory.addContractorOperator(contractor, operator));
        operators.add(EntityFactory.addContractorOperator(contractor, anotherOperator));

        when(propertyDAO.find("admin_email_address")).thenReturn(
                new AppProperty("admin_email_address", "foo@example.com"));
    }

    @Test
    public void testExecute_All() throws Exception {
        cron.execute();
        String report = cron.getReport();
        report = report.replaceAll(" Cron Job at: ([A-Za-z 0-9:]*)", " Cron Job at: TIME_WAS_HERE");
        report = report.replaceAll("\\([0-9]+ millis \\)", "\\(0 millis \\)");
        Approvals.verify(report);
    }

    @Test
    public void testExecute_RunNone() throws Exception {
        when(propertyDAO.getProperty(anyString())).thenReturn("false");
        cron.execute();
        String report = cron.getReport();
        report = report.replaceAll(" Cron Job at: ([A-Za-z 0-9:]*)", " Cron Job at: TIME_WAS_HERE");
        report = report.replaceAll("\\([0-9]+ millis \\)", "\\(0 millis \\)");
        Approvals.verify(report);
    }

    @Test
    public void testSplitPendingAndDeliquentInvoices() throws Exception {
        ContractorAccount deactContractor = EntityFactory.makeContractor();
        ContractorAccount delinContractor = EntityFactory.makeContractor();

        Calendar date = Calendar.getInstance();

        Invoice deactInvoice = new Invoice();
        deactInvoice.setAccount(deactContractor);
        date.add(Calendar.DATE, -1);
        deactInvoice.setDueDate(date.getTime());
        Invoice delinInvoice = new Invoice();
        delinInvoice.setAccount(delinContractor);
        date.add(Calendar.DATE, 6);
        delinInvoice.setDueDate(date.getTime());

        List<Invoice> invoices = new ArrayList<>();
        invoices.add(deactInvoice);
        invoices.add(delinInvoice);

        List<Integer> idList = new ArrayList<>();
        idList.add(deactContractor.getId());
        idList.add(delinContractor.getId());

        when(emailQueueDAO.findContractorsWithRecentEmails(org.mockito.Matchers.anyString(), org.mockito.Matchers.anyInt())).thenReturn(idList);

        Map<ContractorAccount, Integer> map = Whitebox.invokeMethod(cron, "splitPendingAndDeliquentInvoices", invoices);
        assertTrue(map.size() == 0);
    }

    @Test
    public void testRemoveContractorsWithRecentlySentEmail() throws Exception {
        List<ContractorAccount> contractorList = new ArrayList<>();
        contractorList.add(EntityFactory.makeContractor());
        contractorList.add(EntityFactory.makeContractor());
        contractorList.add(EntityFactory.makeContractor());

        List<Integer> idList = new ArrayList<>();
        idList.add(contractorList.get(1).getId()); // get middle contractor

        when(emailQueueDAO.findContractorsWithRecentEmails(org.mockito.Matchers.anyString(), org.mockito.Matchers.anyInt())).thenReturn(idList);
        Whitebox.invokeMethod(cron, "removeContractorsWithRecentlySentEmail", contractorList, 1);
        assertTrue(contractorList.size() == 2);
    }

    @Test
    public void testSendEmailPendingAccounts_SqlHasCountryRestrictionWhenEmailExclusionsExist() throws Exception {
        doTestSqlHasCountryRestriction();
    }

    @Test
    public void testSendEmailPendingAccounts_SqlHasCountryRestrictionWhenNoEmailExclusions() throws Exception {
        @SuppressWarnings("serial")
        List<String> emailExclusionList = new ArrayList<String>() {
            {
                add("123");
                add("124");
            }
        };
        Whitebox.setInternalState(cron, "emailExclusionList", emailExclusionList);
        doTestSqlHasCountryRestriction();
    }

    private void doTestSqlHasCountryRestriction() throws Exception {
        List<ContractorAccount> emptyList = new ArrayList<ContractorAccount>();
        when(contractorAccountDAO.findPendingAccounts(anyString())).thenReturn(emptyList);

        Whitebox.invokeMethod(cron, "sendEmailPendingAccounts");

        ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);

        verify(contractorAccountDAO, times(3)).findPendingAccounts(captor.capture());

        List<String> allSql = captor.getAllValues();

        for (String sql : allSql) {
            assertTrue(sql.contains("a.country IN ('US','CA')"));
        }
    }

    @Test
    public void testSumFlagChanges_EmptyBasicDynaBeanList() throws Exception {
        assertEquals(Integer.valueOf(0), Whitebox.invokeMethod(cron, "sumFlagChanges", (List<BasicDynaBean>) null));
        assertEquals(Integer.valueOf(0), Whitebox.invokeMethod(cron, "sumFlagChanges", new ArrayList<BasicDynaBean>()));
    }

    @Test
    public void testSumFlagChanges_ValidBasicDynaBeanList() throws Exception {
        BasicDynaBean mockDynaBean1 = Mockito.mock(BasicDynaBean.class);
        when(mockDynaBean1.get("changes")).thenReturn(1);
        BasicDynaBean mockDynaBean2 = Mockito.mock(BasicDynaBean.class);
        when(mockDynaBean2.get("changes")).thenReturn(2);

        List<BasicDynaBean> fakes = new ArrayList<BasicDynaBean>();
        fakes.add(mockDynaBean1);
        fakes.add(mockDynaBean2);
        assertEquals(Integer.valueOf(3), Whitebox.invokeMethod(cron, "sumFlagChanges", fakes));
    }

    @Test
    public void testSumFlagChanges_BasicDynaBeanListWithNonIntegerValue() throws Exception {
        BasicDynaBean mockDynaBean1 = Mockito.mock(BasicDynaBean.class);
        when(mockDynaBean1.get("changes")).thenReturn("Fail here!");
        BasicDynaBean mockDynaBean2 = Mockito.mock(BasicDynaBean.class);
        when(mockDynaBean2.get("changes")).thenReturn(2);
        BasicDynaBean mockDynaBean3 = Mockito.mock(BasicDynaBean.class);
        when(mockDynaBean3.get("changes")).thenReturn(null);

        List<BasicDynaBean> fakes = new ArrayList<BasicDynaBean>();
        fakes.add(mockDynaBean1);
        fakes.add(mockDynaBean2);
        fakes.add(mockDynaBean3);
        assertEquals(Integer.valueOf(2), Whitebox.invokeMethod(cron, "sumFlagChanges", fakes));
    }

    @Test
    public void testSumFlagChanges_BasicDynaBeanThrowsException() throws Exception {
        BasicDynaBean mockDynaBean1 = Mockito.mock(BasicDynaBean.class);
        when(mockDynaBean1.get("changes")).thenReturn(new IllegalArgumentException("Forcing an error"));
        BasicDynaBean mockDynaBean2 = Mockito.mock(BasicDynaBean.class);
        when(mockDynaBean2.get("changes")).thenReturn(5);
        List<BasicDynaBean> fakes = new ArrayList<BasicDynaBean>();
        fakes.add(mockDynaBean1);
        fakes.add(mockDynaBean2);
        assertEquals(Integer.valueOf(5), Whitebox.invokeMethod(cron, "sumFlagChanges", fakes));
    }

    @Test
    public void testSendEmailsTo_SingleContractorAccountGetsSingleEmail() throws Exception {
        ContractorAccount cAccount = new ContractorAccount();
        Map<ContractorAccount, Integer> fakeContractors = new TreeMap<ContractorAccount, Integer>();
        fakeContractors.put(cAccount, 48);

        EmailQueue email = new EmailQueue();
        email.setContractorAccount(new ContractorAccount(3));

        when(emailBuilder.build()).thenReturn(email);

        Whitebox.invokeMethod(cron, "sendEmailsTo", fakeContractors);

        verify(emailQueueDAO).save(any(EmailQueue.class));
        verify(noteDAO).save(any(Note.class));
    }

    @Test
    public void testSendInvalidEmailsToBilling_SingleContractorAccountGetsSingleEmail() throws Exception {
        ContractorAccount cAccount = new ContractorAccount();
        Map<ContractorAccount, Integer> fakeContractors = new TreeMap<ContractorAccount, Integer>();
        fakeContractors.put(cAccount, 48);

        EmailQueue email = new EmailQueue();
        email.setContractorAccount(new ContractorAccount(3));

        when(emailBuilder.build()).thenReturn(email);

        Whitebox.invokeMethod(cron, "sendEmailsTo", fakeContractors);

        verify(emailQueueDAO).save(any(EmailQueue.class));
        verify(noteDAO).save(any(Note.class));
    }

    @Test
    public void testRunCRREmailBlast() throws Exception {
        List<ContractorRegistrationRequest> list = new ArrayList<ContractorRegistrationRequest>();
        ContractorRegistrationRequest crr = new ContractorRegistrationRequest();
        crr.setName("test");
        list.add(crr);

        EmailQueue email = new EmailQueue();
        email.setContractorAccount(new ContractorAccount(3));
        // email.setEmailTemplate(any(EmailTemplate.class));
        // when(emailBuilder.build()).thenReturn(email);

        // Whitebox.invokeMethod(cron, "runCRREmailBlast", list, 1, "test");

        // verify(emailQueueDAO).save(any(EmailQueue.class));
        // verify(contractorRegistrationRequestDAO).save(any(ContractorRegistrationRequest.class));
    }

    @Test
    public void testRunAccountEmailBlast() throws Exception {
        List<ContractorAccount> list = new ArrayList<ContractorAccount>();
        ContractorAccount cAccount = new ContractorAccount(1);
        list.add(cAccount);

        EmailQueue email = new EmailQueue();
        email.setContractorAccount(new ContractorAccount(3));

        // when(emailBuilder.build()).thenReturn(email);

        // Whitebox.invokeMethod(cron, "runAccountEmailBlast", list, 1, "test");
        // verify(emailQueueDAO).save(any(EmailQueue.class));
        // verify(contractoAccountDAO).save(any(ContractorAccount.class));
    }

    @Test
    public void testDeactivatePendingAccounts() throws Exception {
        ContractorAccount cAccount = new ContractorAccount(1);
        ContractorAccount cAccount2 = new ContractorAccount(2);
        cAccount.setStatus(AccountStatus.Pending);
        cAccount2.setStatus(AccountStatus.Pending);
        List<ContractorAccount> cList = new ArrayList<>();
        cList.add(cAccount);
        cList.add(cAccount2);

        when(contractorAccountDAO.findPendingAccountsToMoveToDeclinedStatus()).thenReturn(cList);

        cron.move90DayPendingAccountsToDeclinedStatus();

        verify(accountStatusChanges).declineContractor(cAccount, permissions,
                AccountStatusChanges.DID_NOT_COMPLETE_PICS_PROCESS_REASON,
                AccountStatusChanges.NOTE_DID_NOT_COMPLETE_PICS_PROCESS_REASON);
        verify(accountStatusChanges).declineContractor(cAccount2, permissions,
                AccountStatusChanges.DID_NOT_COMPLETE_PICS_PROCESS_REASON,
                AccountStatusChanges.NOTE_DID_NOT_COMPLETE_PICS_PROCESS_REASON);
    }

    @Test
    public void testDeactivatePendingAccounts_emptyList() {
        List<ContractorAccount> cList = new ArrayList<>();
        when(contractorAccountDAO.findPendingAccountsToMoveToDeclinedStatus()).thenReturn(cList);
        verify(accountStatusChanges, never()).deactivateContractor(any(ContractorAccount.class),
                any(Permissions.class), anyString(), anyString());
    }

    @Test
    public void testDeactivateNonRenewalAccounts() throws Exception {
        String where = "a.status = 'Active' AND a.renew = 0 AND paymentExpires < NOW()";
        List<ContractorAccount> contractors = buildMockContractorList();
        when(contractorAccountDAO.findWhere(where)).thenReturn(contractors);

        Whitebox.invokeMethod(cron, "deactivateNonRenewalAccounts");

        for (ContractorAccount contractor : contractors) {
            verifyContractor(contractor, contractor.getAccountLevel().isBidOnly());
        }
    }

    private List<ContractorAccount> buildMockContractorList() {
        List<ContractorAccount> contractors = new ArrayList<ContractorAccount>();
        contractors.add(buildMockContractorAccount(1, false));
        contractors.add(buildMockContractorAccount(2, true));

        return contractors;
    }

    private ContractorAccount buildMockContractorAccount(int id, boolean isBidOnly) {
        ContractorAccount contractor = Mockito.mock(ContractorAccount.class);
        when(contractor.getId()).thenReturn(id);

        if (isBidOnly) {
            when(contractor.getAccountLevel()).thenReturn(AccountLevel.BidOnly);
        } else {
            when(contractor.getAccountLevel()).thenReturn(AccountLevel.Full);
        }

        return contractor;
    }

    private void verifyContractor(ContractorAccount contractor, boolean isBidOnly) {
        String reason = isBidOnly ? AccountStatusChanges.BID_ONLY_ACCOUNT_REASON
                : AccountStatusChanges.DEACTIVATED_NON_RENEWAL_ACCOUNT_REASON;

        verify(billingService, times(1)).syncBalance(contractor);
        verify(contractor, times(1)).setAuditColumns(Cron.system);
        verify(accountStatusChanges, times(1)).deactivateContractor(contractor, permissions, reason,
                "Automatically inactivating account based on expired membership");
    }

    @Test
    public void testCalculateLateFeeFor() {
        BigDecimal invoiceFeeAmount = new BigDecimal(100);
        BigDecimal expectedLateFee = new BigDecimal(Cron.MINIMUM_LATE_FEE);
        when(mockInvoice.getTotalAmount()).thenReturn(invoiceFeeAmount);
        BigDecimal actualLateFee = cron.calculateLateFeeFor(mockInvoice);
        assertEquals(expectedLateFee, actualLateFee);
    }

    @Test
    public void createLateFeeInvoiceItem() {
        when(invoiceFeeDAO.findByNumberOfOperatorsAndClass(FeeClass.LateFee, 0)).thenReturn(mockInvoiceFee);
        when(mockInvoiceFee.getFeeClass()).thenReturn(FeeClass.LateFee);
        when(mockInvoice.getAccount()).thenReturn(mockContractorAccount);
        BigDecimal invoiceFeeAmount = new BigDecimal(100);
        when(mockInvoice.getTotalAmount()).thenReturn(invoiceFeeAmount);
        BigDecimal lateFee = cron.calculateLateFeeFor(mockInvoice);
        InvoiceItem lateFeeInvoiceItem = cron.createLateFeeInvoiceItem(mockInvoice, lateFee);
        assertEquals(FeeClass.LateFee, lateFeeInvoiceItem.getInvoiceFee().getFeeClass());
        assertEquals(lateFee, lateFeeInvoiceItem.getAmount());
    }

    @Test
    public void testGenerateLateFeeInvoice() {
        when(invoiceFeeDAO.findByNumberOfOperatorsAndClass(FeeClass.LateFee, 0)).thenReturn(mockInvoiceFee);
        when(mockInvoiceFee.getFeeClass()).thenReturn(FeeClass.LateFee);
        when(mockInvoice.getAccount()).thenReturn(mockContractorAccount);
        Date invoiceDueDate = new Date();
        when(mockInvoice.getDueDate()).thenReturn(invoiceDueDate);
        BigDecimal invoiceFeeAmount = new BigDecimal(100);
        when(mockInvoice.getTotalAmount()).thenReturn(invoiceFeeAmount);
        BigDecimal lateFee = cron.calculateLateFeeFor(mockInvoice);
        InvoiceItem lateFeeInvoiceItem = cron.createLateFeeInvoiceItem(mockInvoice, lateFee);
        Invoice lateInvoice = cron.generateLateFeeInvoice(mockInvoice, lateFeeInvoiceItem);

        assertEquals(mockInvoice.getAccount(), lateInvoice.getAccount());
        assertEquals(lateFeeInvoiceItem, lateInvoice.getItems().get(0));
        assertEquals(lateFee, lateInvoice.getTotalAmount());
        assertEquals(InvoiceType.LateFee, lateInvoice.getInvoiceType());
        assertEquals(FeeClass.LateFee, lateInvoice.getItems().get(0).getInvoiceFee().getFeeClass());
        assertEquals(invoiceDueDate.getDay(), lateInvoice.getDueDate().getDay());
        assertEquals(invoiceDueDate.getMonth(), lateInvoice.getDueDate().getMonth());
        assertEquals(invoiceDueDate.getYear(), lateInvoice.getDueDate().getYear());
    }
}
