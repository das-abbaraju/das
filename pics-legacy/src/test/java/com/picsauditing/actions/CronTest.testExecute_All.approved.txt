Running Cron Job on Server: null

Address: null

Cron Job initiated by: null

Starting Cron Job at: TIME_WAS_HERE


Starting Cron.RunStep.AuditBuilder_addAuditRenewals
	SUCCESS...(0 millis )

Starting Running Huntsman EBIX Support...
	SUCCESS...(0 millis )

Starting Recalculating all the categories for Audits...
	SUCCESS...(0 millis )

Starting Starting Indexer...
	SUCCESS...(0 millis )

Starting Sending emails to contractors pending...
	SUCCESS...(0 millis )

Starting Sending emails to registration requests...
	SUCCESS...(0 millis )

Starting Sending email about upcoming implementation audits...
	SUCCESS...(0 millis )

Starting Adding Late Fee to Delinquent Contractor Invoices ...
	No invoices require a late fee

SUCCESS...(0 millis )

Starting Bump Dead Accounts that still have balances...
	SUCCESS...(0 millis )

Starting Inactivating Accounts via Billing Status...
	SUCCESS...(0 millis )

Starting Sending Email to Delinquent Contractors ...
	SUCCESS...(0 millis )

Starting Sending No Action Email to Bid Only Accounts ...
	SUCCESS...(0 millis )

Starting Stamping Notes and Expiring overall Forced Flags and Individual Data Overrides...
	SUCCESS...(0 millis )

Starting Expiring Flag Changes...
	SUCCESS...(0 millis )

Starting Refresh Report Suggestions...
	SUCCESS...(0 millis )

Starting Emailing Flag Change Reports...
	SUCCESS...(0 millis )

Starting Checking System Status...
	SUCCESS...(0 millis )

Starting Checking Registration Requests Hold Dates...
	SUCCESS...(0 millis )

Starting Moving pending accounts past 90 days to declined
	SUCCESS...(0 millis )




Completed Cron Job at: TIME_WAS_HERE